File: Applied Course/5.Advanced Data Structure and Algorithms/3.LRU Cache.cpp
#include<bits/stdc++.h>
using namespace std;
struct Node {
    int key;
    Node* prev = nullptr;
    Node* next = nullptr;
    Node(int key) : key(key) {};
};

class DoubleLinkedList {
public:
    Node* head = nullptr;
    Node* tail = nullptr;
    Node* append(int key) {
        // append to linked list
        Node* n = new Node(key);
        // if tail is null then head should be null so just set both to a new Node
        if(!tail) {
            head = tail = n;
        }
        // else if tail is not null so point tail to new node and set tail to be n
        else {
            tail->next = n;
            n->prev = tail;
            tail = n;
        }
        return n;
    }
    void moveToEnd(Node* n) {
        assert(head != nullptr && tail != nullptr);
        if(n == tail) return;
        if(n == head) {
            tail->next = n;
            n->prev = tail;
            tail = n;
            head = n->next;
            n->next = nullptr;
        } else { // in between
            n->prev->next = n->next;
            n->next->prev = n->prev;
            tail->next = n;
            n->prev = tail;
            tail = n;
            n->next = nullptr;
        }
    }
    // deletes the head. head is now second element 
    // returns key of head
    int deleteHead() {
        assert(head != nullptr);
        Node* newHead = head->next;
        int delKey = head->key;
        delete head;
        head = newHead;
        return delKey;
    } 
};

class LRUCache {
private:
    DoubleLinkedList dll;
    unordered_map<int,pair<int, Node*>> m;
    int capacity;
public:
    LRUCache(int _capacity) {
        assert(_capacity > 0);
        capacity = _capacity;
    }
    
    int get(int key) {
        auto it = m.find(key);
        if(it == m.end()) return -1;
        dll.moveToEnd(it->second.second);
        return it->second.first;
    }
    
    void put(int key, int value) {
        auto it = m.find(key); 
        if(it == m.end()) {
            Node* n = dll.append(key);
            m[key] = {value, n};
            // if size of map exceeds capacity erase lru item from map and dll 
            if(m.size() > capacity) m.erase(dll.deleteHead());
        } else {
            it->second.first = value;
            dll.moveToEnd(it->second.second);
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
int main()
{
    // https://leetcode.com/problems/lru-cache/
    return 0;
}

File: Applied Course/5.Advanced Data Structure and Algorithms/3.LRU Cache.py
# https://leetcode.com/problems/lru-cache/
class DLLNode:
    def __init__(self):
        self.key=0
        self.value=0
        self.prev=None
        self.next=None
        
class LRUCache:
    
    def _remove_node(self,node):
        prev=node.prev 
        next=node.next 
        prev.next=next 
        next.prev=prev 
    
    def _add_node(self,node):
        node.prev=self.head
        node.next=self.head.next
        self.head.next.prev=node
        self.head.next=node
        
    def _remove_tail_node(self):
        res=self.tail.prev
        self._remove_node(res)
        return res
        
    def _move_to_head(self,node):
        self._remove_node(node)
        self._add_node(node)

    def __init__(self, capacity: int):
        self.cache={}
        self.size=0
        self.capacity=capacity
        self.head,self.tail=DLLNode(),DLLNode()
        self.head.next=self.tail
        self.tail.prev=self.head

    def get(self, key: int) -> int:
        node=self.cache.get(key,None)        
        if not node:
            return -1
        # move the item to head
        self._move_to_head(node)
        return node.value

    def put(self, key: int, value: int) -> None:
        node=self.cache.get(key)
        if not node:
            newNode=DLLNode()
            newNode.key=key
            newNode.value=value
            self.cache[key]=newNode
            self._add_node(newNode)
            self.size+=1
            if self.size>self.capacity:
                # remove tail node of the cache (dll)
                tail=self._remove_tail_node()
                del self.cache[tail.key]
                self.size-=1
        else:
            node.value=value
            self._move_to_head(node)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)